arch/xtensa/src/esp32s2/esp32s2_tim.c

```diff
@@ -0,0 +1,981 @@
+static void esp32s2_tim_setpre(FAR struct esp32s2_tim_dev_s *dev,
+                               uint16_t pre)
+{
+  struct esp32s2_tim_priv_s *priv = (FAR struct esp32s2_tim_priv_s *)dev;
+  uint32_t mask = (uint32_t)pre << TIMG_T0_DIVIDER_S;
+
+  DEBUGASSERT(dev);
+
+  if (priv->tid == TIMER0)
+    {
+      modifyreg32(TIMG_T0CONFIG_REG(priv->gid), TIMG_T0_DIVIDER_M, mask);
+    }
+  else
+    {
+      modifyreg32(TIMG_T1CONFIG_REG(priv->gid), TIMG_T1_DIVIDER_M, mask);
+    }
+}
+
+/****************************************************************************
+ * Name: esp32s2_tim_getcounter
+ *
+ * Description:
+ *   Get the current counter value.
+ *
+ * Parameters:
+ *   dev           - Pointer to the timer driver struct.
+ *   value         - A pointer to a variable to store the current read
+ *                   value from counter.
+ *
+ ****************************************************************************/
+
+static void esp32s2_tim_getcounter(FAR struct esp32s2_tim_dev_s *dev,
+                                   uint64_t *value)
+{
+  uint32_t value_32;
+  struct esp32s2_tim_priv_s *priv = (FAR struct esp32s2_tim_priv_s *)dev;
+
+  DEBUGASSERT(dev);
+
+  *value = 0;
+
+  if (priv->tid == TIMER0)
+    {
+      /* Dummy value (0 or 1) to latch the counter value to read it */
+
+      putreg32(BIT(0), TIMG_T0UPDATE_REG(priv->gid));
+
+      /* Read value */
+
+      value_32  = getreg32(TIMG_T0HI_REG(priv->gid)); /* High 32 bits */
+      *value   |= (uint64_t)value_32;
+      *value  <<= SHIFT_32;
+      value_32  = getreg32(TIMG_T0LO_REG(priv->gid)); /* Low 32 bits */
+      *value   |= (uint64_t)value_32;
+    }
+  else
+    {
+      /* Dummy value (0 or 1) to latch the counter value to read it */
+
+      putreg32(BIT(0), TIMG_T1UPDATE_REG(priv->gid));
+
+      /* Read value */
+
+      value_32  = getreg32(TIMG_T1HI_REG(priv->gid)); /* High 32 bits */
+      *value   |= (uint64_t)value_32;
+      *value  <<= SHIFT_32;
+      value_32  = getreg32(TIMG_T1LO_REG(priv->gid)); /* Low 32 bits */
+      *value   |= (uint64_t)value_32;
+    }
+}
+
+/****************************************************************************
+ * Name: esp32s2_tim_setcounter
+ *
+ * Description:
+ *   Set the value to be loaded to the counter.
+ *   If you want the counter to be loaded at an alarm, enable the alarm and
+ *   the auto-reload before.
+ *   If you want the counter to be loaded instantly, call
+ *   esp32s2_tim_reload_now() after this function.
+ *
+ * Parameters:
+ *   dev           - Pointer to the timer driver struct.
+ *   value         - The value to be loaded the counter.
+ *
+ ****************************************************************************/
+
+static void esp32s2_tim_setcounter(FAR struct esp32s2_tim_dev_s *dev,
+                                  uint64_t value)
+{
+  uint64_t low_64 = value & UINT32_MAX;
+  uint64_t high_64 = (value >> SHIFT_32);
+  struct esp32s2_tim_priv_s *priv = (FAR struct esp32s2_tim_priv_s *)dev;
```

Comment:
Please move the struct to the first line inside the function, normally structs and complex types come first

Reply:
All fixed.